### PR TITLE
csi: Implement PV Resize

### LINKED ISSUE #45: RFE: PV resize support (through CSI module)
Decide and design possible resize feature.

Will add more details soon. Ping here if you are interested to pick it up

### PR DESCRIPTION
**Pull request by vatsa287**
This PR introduces support to expand PV volumes. This support is added through adding a new side-car container 'csi-resizer' within 'kadalu-provisioner' image. Implements controller_expand_volume() as per [CSI-spec](https://github.com/container-storage-interface/spec/blob/master/spec.md). 

To expand volume :

1. Set 'allowVolumeExpansion' to 'true' in storage class.
2. Terminate the pod using the volume. [Only for VM attached or these types of volumes.](https://github.com/kubernetes/kubernetes/issues/68427).
3. Patch the pvc to be expanded to new size
    Example :
   ```kubectl patch pvc <pvc-name> -p '{"spec":{"resources":{"requests":{"storage":"70Gi"}}}}}' ```

Notice that size and quota have been updated through:

```kubectl describe pvc <pvc-name>```
and
```kubectl logs -nkadalu server-*  quotad```

> Note: "*" is the storage-pool of which "pvc-name" is part of and found using "kubectl get pods -nkadalu".
> Shrinking of size is not supported. [Read here.](https://kubernetes.io/blog/2018/07/12/resizing-persistent-volumes-using-kubernetes/)

Fixes: #45 
Signed-off-by: Shree Vatsa N <vatsa@kadalu.io>

### CODE REVIEW

**Comment by vatsa287** [comment]
Just trying to implement this feature directly without checking for logic as mentioned in #45 as of now for debugging. 
Getting this after editing a RWO PVC. The ControllerExpandVolume is not being called. Will try to complete this after exams in Feb.
```
 Warning ExternalExpanding 2m19s volume_expand Ignoring the PVC: didn't find a plugin capable of expanding the volume; waiting for an external controller to process this PVC.
```

**Comment by aravindavk** [comment]
> Just trying to implement this feature directly without checking for logic as mentioned in #45 as of now for debugging.
> Getting this after editing a RWO PVC. The ControllerExpandVolume is not being called. Will try to complete this after exams in Feb.
> 
> ```
>  Warning ExternalExpanding 2m19s volume_expand Ignoring the PVC: didn't find a plugin capable of expanding the volume; waiting for an external controller to process this PVC.
> ```

One more sidecar container is required to handle Resize request. Check this https://github.com/kubernetes-csi/external-resizer

**Comment by vatsa287** [comment]
@aravindavk Thanks. I went through docs and added csi-resizer as a container inside csi-provisioner as it is done for attacher, kadalu-logging and gave permissions for it to watch the pod.

**Comment by amarts** [comment]
```
statefulset.apps/kadalu-csi-provisioner created Error from server (Forbidden): error when creating "/kadalu/templates/csi.yaml": clusterroles.rbac.authorization.k8s.io "kadalu-csi-provisioner" is forbidden: user "system:serviceaccount:kadalu:kadalu-operator" (groups=["system:serviceaccounts" "system:serviceaccounts:kadalu" "system:authenticated"]) is attempting to grant RBAC permissions not currently held:
{APIGroups:[""], Resources:["pods/exec"], Verbs:["create" "get" "list" "watch" "update" "delete" "patch"]}
{APIGroups:[""], Resources:["pods/log"], Verbs:["create" "get" "list" "watch" "update" "delete" "patch"]}
Error from server (NotFound): error when creating "/kadalu/templates/csi.yaml": clusterroles.rbac.authorization.k8s.io "kadalu-csi-provisioner" not found
Error from server (Forbidden): error when creating "/kadalu/templates/csi.yaml": roles.rbac.authorization.k8s.io "kadalu-csi-provisioner" is forbidden: user "system:serviceaccount:kadalu:kadalu-operator" (groups=["system:serviceaccounts" "system:serviceaccounts:kadalu" "system:authenticated"]) is attempting to grant RBAC permissions not currently held:
{APIGroups:[""], Resources:["pods/exec"], Verbs:["create" "get" "list" "watch" "update" "delete" "patch"]}
{APIGroups:[""], Resources:["pods/log"], Verbs:["create" "get" "list" "watch" "update" "delete" "patch"]}
Error from server (NotFound): error when creating "/kadalu/templates/csi.yaml": roles.rbac.authorization.k8s.io "kadalu-csi-provisioner" not found
```

File: csi/nodeserver.py [hunk 1]
```diff
@@ -107,3 +107,38 @@ def NodeGetInfo(self, request, context):
...
+        logging.debug(logf(
+            "Received the expand request",
+            volume=request.volume_id,
+            voltype=voltype,
+            hostvol=hostvol,
+            pvpath=pvpath,
+            pvtype=pvtype
+        ))
+
+        expand_volume(request.target_path)
```
Comment (amarts) [inline csi/nodeserver.py]: Is it called for RWX volume? if yes, then we don't need to do anything at all, as quota update is good enough. Only for Block image, we need to do `grow_xfs`.

**Review by amarts — COMMENTED** [review]
Can you add a test for volume expansion? That way we are sure to test it everytime?

I guess just calling volume size change in Replica3 type PVC yaml and then calling the same test again (without pv delete) is good IMO.

**Comment by vatsa287** [comment]
@amarts 
```
Conditions:
  Type       Status  LastProbeTime                     LastTransitionTime                Reason  Message
  ----       ------  -----------------                 ------------------                ------  -------
  Resizing   True    Mon, 01 Jan 0001 00:00:00 +0000   Tue, 05 Jan 2021 18:39:49 +0530           
Events:
  Type    Reason    Age   From                     Message
  ----    ------    ----  ----                     -------
  Normal  Resizing  29s   external-resizer kadalu  External resizer is resizing volume pvc-27cf4df9-879f-4c69-9487-23da60fff5ea
``` 
Looks like the operator was missing "persistentvolumeclaims/status" type... I'll try to implement logic by next updates

File: csi/controllerserver.py [hunk 2]
```diff
@@ -255,6 +258,118 @@ def ControllerGetCapabilities(self, request, context):
...
+            context.set_code(grpc.StatusCode.RESOURCE_EXHAUSTED)
+            return csi_pb2.CreateVolumeResponse()
+
+        mntdir = os.path.join(HOSTVOL_MOUNTDIR, hostvol)
+
+        if request.vol_capability.BlockVolume:
+            logging.info(logf(
+                "Volume is of type 'block'"
+            ))
+            expand_volume(mntdir)
```
Comment (vatsa287) [inline csi/controllerserver.py:342]: Is this the way of handling BlockVolume

Comment (vatsa287) [inline csi/controllerserver.py]: Is this the way of handling BlockVolume?

File: csi/controllerserver.py [hunk 3]
```diff
@@ -255,6 +258,118 @@ def ControllerGetCapabilities(self, request, context):
...
+            "Filters applied to choose storage",
+            **filters
+        ))
+
+        # UID is stored at the time of installation in configmap.
+        uid = None
+        with open(os.path.join(VOLINFO_DIR, "uid")) as uid_file:
+            uid = uid_file.read()
+
+        host_volumes = get_pv_hosting_volumes(filters)
```
Comment (vatsa287) [inline csi/controllerserver.py]: What is the right way of getting the host_vol.. Since 'filters' is not available through ControllerExpandVolume,   request.parameters is not supported. Only volume_id, capacity_range & volume_capability is supported.

Comment (amarts) [reply csi/controllerserver.py]: dump what you get in `request` and `context`... that should give you enough hints.

File: csi/controllerserver.py (hunk 2 above)
Comment (amarts) [reply csi/controllerserver.py]: Almost right! but I believe we will handle both VIRTBLOCK and BlockVolume almost similarly! Hence, doing the `expand_volume()` after `update_virtblock_volume()` is better. (in the same code block).

Comment (vatsa287) [reply csi/controllerserver.py]: sure

File: csi/controllerserver.py [hunk 4]
```diff
@@ -255,6 +258,133 @@ def ControllerGetCapabilities(self, request, context):
...
+        # subtracted from storage-pool
+        # sizechange = expansion_requested_pvsize - existing_pvsize
+        # update_free_size(hostvol, request.name, -sizechange)
+
+        # if not hostvoltype:
+        #    hostvoltype = "unknown"
+
+        # send_analytics_tracker("pvc-%s" % hostvoltype, uid)
+        return csi_pb2.ControllerExpandVolumeResponse({
+                "capacity_bytes": expansion_requested_pvsize,
```
Comment (vatsa287) [inline csi/controllerserver.py]: I think the response message is wrong, I'm not getting the right format. Below is the prescribed format
```
message ControllerExpandVolumeResponse {
  // Capacity of volume after expansion. This field is REQUIRED.
  int64 capacity_bytes = 1;

  // Whether node expansion is required for the volume. When true
  // the CO MUST make NodeExpandVolume RPC call on the node. This field
  // is REQUIRED.
  bool node_expansion_required = 2;
}
``` 
I'm getting error 
```
 return csi_pb2.ControllerExpandVolumeResponse({
TypeError: No positional arguments allowed
```

Comment (amarts) [reply csi/controllerserver.py]: Nope, don't pass anything here. That is handled by csi_pb2.py file.

File: csi/controllerserver.py (hunk 3 above)
Comment (vatsa287) [reply csi/controllerserver.py]: I got it through search_volume(request.volume_id)

**Comment by amarts** [comment]
fix the pylint too while at it.. also submit your changes as separate commit from the next fix, so we can continue to remember what was tried, and you can always squash merge.

**Comment by vatsa287** [comment]
> fix the pylint too while at it.. also submit your changes as separate commit from the next fix, so we can continue to remember what was tried, and you can always squash merge.

Sure!

File: templates/csi.yaml.j2 [hunk 5]
```diff
@@ -328,6 +404,21 @@ spec:
           volumeMounts:
             - name: varlog
               mountPath: "/var/log/gluster"
+        - name: csi-resizer
+          image: k8s.gcr.io/sig-storage/csi-resizer:v1.0.1
```
Comment (amarts) [inline templates/csi.yaml.j2]: can you use `docker.io/raspbernetes/csi-external-resizer:1.0.0` here instead of above? That way, we don't break arm64 and arm/v7 support.

Comment (vatsa287) [reply templates/csi.yaml.j2]: done!

File: templates/storageclass-kadalu.yaml.j2 [hunk 6]
```diff
@@ -5,3 +5,4 @@ apiVersion: storage.k8s.io/v1
 metadata:
   name: kadalu
 provisioner: kadalu
+allowVolumeExpansion: true
```
Comment (amarts) [inline templates/storageclass-kadalu.yaml.j2:8]: Should we add this option in 'external' storage storage class too? (Technically, we are already supporting it).

I am OK to do that as a separate patch, as we do it with CLI and also in documentation, but not in the code anywhere.

**Review by amarts — APPROVED** [review]
Looks like this is ready to get merged, let me know how you tested, and how you plan to add it to regular tests.

**Comment by amarts** [comment]
@vatsa287 Please check why below steps are necessary:

> Set 'allowVolumeExpansion' to 'true' in storage class.
> Terminate the pod using the volume.
> Then patch the PVC.

glusterfs supports online volume expansion. I feel we can support online volume expansion, and once this PR is merged, please look at what it takes to enable `onlineVolumeExpansion` feature. Stopping the pod is not ideal for our users.

File: templates/storageclass-kadalu.yaml.j2 (hunk 6 above)
Comment (vatsa287) [reply templates/storageclass-kadalu.yaml.j2:8]: ### TEST

Created sample-pvc with 100Mi as claim with allowVolumeExpansion: true in stroage class
```
vatsa@vatsa-VirtualBox:~/kadalu$ sudo kubectl describe pvc pv1
Name:          pv1
Namespace:     default
StorageClass:  kadalu.replica1
Status:        Bound
Volume:        pvc-10cbbc2a-8591-4db6-a6ff-11326ed4cd2a
Labels:        <none>
Annotations:   kubectl.kubernetes.io/last-applied-configuration:
                 {"apiVersion":"v1","kind":"PersistentVolumeClaim","metadata":{"annotations":{},"name":"pv1","namespace":"default"},"spec":{"accessModes":[...
               pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               volume.beta.kubernetes.io/storage-provisioner: kadalu
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      100Mi
Access Modes:  RWO
VolumeMode:    Filesystem
Mounted By:    <none>
Events:
  Type    Reason                 Age   From                                                                  Message
  ----    ------                 ----  ----                                                                  -------
  Normal  ExternalProvisioning   11s   persistentvolume-controller                                           waiting for a volume to be created, either by external provisioner "kadalu" or manually created by system administrator
  Normal  Provisioning           11s   kadalu_kadalu-csi-provisioner-0_cd9bf8b4-7f11-42f3-9a11-57afae19625b  External provisioner is provisioning volume for claim "default/pv1"
  Normal  ProvisioningSucceeded  8s    kadalu_kadalu-csi-provisioner-0_cd9bf8b4-7f11-42f3-9a11-57afae19625b  Successfully provisioned volume pvc-10cbbc2a-8591-4db6-a6ff-11326ed4cd2a
```
Then patched for higher claim of 200Mi

```
vatsa@vatsa-VirtualBox:~/kadalu$ sudo kubectl patch pvc pv1 -p '{"spec":{"resources":{"requests":{"storage":"200Mi"}}}}}'
persistentvolumeclaim/pv1 patched
```
Logs from describe, csi-resizer and quotad
```
vatsa@vatsa-VirtualBox:~/kadalu$ sudo kubectl describe pvc pv1
Name:          pv1
Namespace:     default
StorageClass:  kadalu.replica1
Status:        Bound
Volume:        pvc-10cbbc2a-8591-4db6-a6ff-11326ed4cd2a
Labels:        <none>
Annotations:   kubectl.kubernetes.io/last-applied-configuration:
                 {"apiVersion":"v1","kind":"PersistentVolumeClaim","metadata":{"annotations":{},"name":"pv1","namespace":"default"},"spec":{"accessModes":[...
               pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               volume.beta.kubernetes.io/storage-provisioner: kadalu
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      200Mi
Access Modes:  RWO
VolumeMode:    Filesystem
Mounted By:    pod1
Events:
  Type     Reason                  Age    From                                                                  Message
  ----     ------                  ----   ----                                                                  -------
  Normal   ExternalProvisioning    4m35s  persistentvolume-controller                                           waiting for a volume to be created, either by external provisioner "kadalu" or manually created by system administrator
  Normal   Provisioning            4m35s  kadalu_kadalu-csi-provisioner-0_cd9bf8b4-7f11-42f3-9a11-57afae19625b  External provisioner is provisioning volume for claim "default/pv1"
  Normal   ProvisioningSucceeded   4m32s  kadalu_kadalu-csi-provisioner-0_cd9bf8b4-7f11-42f3-9a11-57afae19625b  Successfully provisioned volume pvc-10cbbc2a-8591-4db6-a6ff-11326ed4cd2a
  Warning  ExternalExpanding       80s    volume_expand                                                         Ignoring the PVC: didn't find a plugin capable of expanding the volume; waiting for an external controller to process this PVC.
  Normal   Resizing                80s    external-resizer kadalu                                               External resizer is resizing volume pvc-10cbbc2a-8591-4db6-a6ff-11326ed4cd2a
  Normal   VolumeResizeSuccessful  77s    external-resizer kadalu                                               Resize volume succeeded
```
```
vatsa@vatsa-VirtualBox:~/kadalu$ sudo kubectl logs -nkadalu kadalu-csi-provisioner-0 csi-resizer
I0114 14:45:03.136339       1 main.go:79] Version : v1.0.0-0-gb22717d-dirty
I0114 14:45:03.225380       1 connection.go:153] Connecting to unix:///var/lib/csi/sockets/pluginproxy/csi.sock
I0114 14:45:03.275644       1 common.go:111] Probing CSI driver for readiness
W0114 14:45:03.278921       1 metrics.go:142] metrics endpoint will not be started because `metrics-address` was not specified.
I0114 14:45:03.291063       1 controller.go:117] Register Pod informer for resizer kadalu
I0114 14:45:03.302767       1 controller.go:241] Starting external resizer kadalu
I0114 14:53:34.598992       1 event.go:282] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"pv1", UID:"10cbbc2a-8591-4db6-a6ff-11326ed4cd2a", APIVersion:"v1", ResourceVersion:"2865496", FieldPath:""}): type: 'Normal' reason: 'Resizing' External resizer is resizing volume pvc-10cbbc2a-8591-4db6-a6ff-11326ed4cd2a
I0114 14:53:37.506997       1 event.go:282] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"pv1", UID:"10cbbc2a-8591-4db6-a6ff-11326ed4cd2a", APIVersion:"v1", ResourceVersion:"2865496", FieldPath:""}): type: 'Normal' reason: 'VolumeResizeSuccessful' Resize volume succeeded
```
```
vatsa@vatsa-VirtualBox:~/kadalu$ sudo kubectl logs -nkadalu server-storage-pool-1-0-0 quotad
+ pid=0
+ cmd=/usr/bin/python3
+ script=/kadalu/server.py
+ trap 'kill ${!}; term_handler' SIGTERM
+ pid=6
+ true
+ wait 7
+ tail -f /dev/null
+ /usr/bin/python3 /kadalu/server.py
[2021-01-14 14:50:22,656] INFO [quotad - 105:handle_quota] - Quota set for size	 size=104857600
[2021-01-14 14:53:37,227] INFO [quotad - 105:handle_quota] - Quota set for size	 size=209715200
```

**Comment by vatsa287** [comment]
### TEST

Created sample-pvc with 100Mi as claim with allowVolumeExpansion: true in storage class
```
vatsa@vatsa-VirtualBox:~/kadalu$ sudo kubectl describe pvc pv1
Name:          pv1
Namespace:     default
StorageClass:  kadalu.replica1
Status:        Bound
Volume:        pvc-10cbbc2a-8591-4db6-a6ff-11326ed4cd2a
Labels:        <none>
Annotations:   kubectl.kubernetes.io/last-applied-configuration:
                 {"apiVersion":"v1","kind":"PersistentVolumeClaim","metadata":{"annotations":{},"name":"pv1","namespace":"default"},"spec":{"accessModes":[...
               pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               volume.beta.kubernetes.io/storage-provisioner: kadalu
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      100Mi
Access Modes:  RWO
VolumeMode:    Filesystem
Mounted By:    <none>
Events:
  Type    Reason                 Age   From                                                                  Message
  ----    ------                 ----  ----                                                                  -------
  Normal  ExternalProvisioning   11s   persistentvolume-controller                                           waiting for a volume to be created, either by external provisioner "kadalu" or manually created by system administrator
  Normal  Provisioning           11s   kadalu_kadalu-csi-provisioner-0_cd9bf8b4-7f11-42f3-9a11-57afae19625b  External provisioner is provisioning volume for claim "default/pv1"
  Normal  ProvisioningSucceeded  8s    kadalu_kadalu-csi-provisioner-0_cd9bf8b4-7f11-42f3-9a11-57afae19625b  Successfully provisioned volume pvc-10cbbc2a-8591-4db6-a6ff-11326ed4cd2a
```
Then patched for higher claim of 200Mi

```
vatsa@vatsa-VirtualBox:~/kadalu$ sudo kubectl patch pvc pv1 -p '{"spec":{"resources":{"requests":{"storage":"200Mi"}}}}}'
persistentvolumeclaim/pv1 patched
```
Logs from describe, csi-resizer and quotad
```
vatsa@vatsa-VirtualBox:~/kadalu$ sudo kubectl describe pvc pv1
Name:          pv1
Namespace:     default
StorageClass:  kadalu.replica1
Status:        Bound
Volume:        pvc-10cbbc2a-8591-4db6-a6ff-11326ed4cd2a
Labels:        <none>
Annotations:   kubectl.kubernetes.io/last-applied-configuration:
                 {"apiVersion":"v1","kind":"PersistentVolumeClaim","metadata":{"annotations":{},"name":"pv1","namespace":"default"},"spec":{"accessModes":[...
               pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               volume.beta.kubernetes.io/storage-provisioner: kadalu
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      200Mi
Access Modes:  RWO
VolumeMode:    Filesystem
Mounted By:    pod1
Events:
  Type     Reason                  Age    From                                                                  Message
  ----     ------                  ----   ----                                                                  -------
  Normal   ExternalProvisioning    4m35s  persistentvolume-controller                                           waiting for a volume to be created, either by external provisioner "kadalu" or manually created by system administrator
  Normal   Provisioning            4m35s  kadalu_kadalu-csi-provisioner-0_cd9bf8b4-7f11-42f3-9a11-57afae19625b  External provisioner is provisioning volume for claim "default/pv1"
  Normal   ProvisioningSucceeded   4m32s  kadalu_kadalu-csi-provisioner-0_cd9bf8b4-7f11-42f3-9a11-57afae19625b  Successfully provisioned volume pvc-10cbbc2a-8591-4db6-a6ff-11326ed4cd2a
  Warning  ExternalExpanding       80s    volume_expand                                                         Ignoring the PVC: didn't find a plugin capable of expanding the volume; waiting for an external controller to process this PVC.
  Normal   Resizing                80s    external-resizer kadalu                                               External resizer is resizing volume pvc-10cbbc2a-8591-4db6-a6ff-11326ed4cd2a
  Normal   VolumeResizeSuccessful  77s    external-resizer kadalu                                               Resize volume succeeded
```
```
vatsa@vatsa-VirtualBox:~/kadalu$ sudo kubectl logs -nkadalu kadalu-csi-provisioner-0 csi-resizer
I0114 14:45:03.136339       1 main.go:79] Version : v1.0.0-0-gb22717d-dirty
I0114 14:45:03.225380       1 connection.go:153] Connecting to unix:///var/lib/csi/sockets/pluginproxy/csi.sock
I0114 14:45:03.275644       1 common.go:111] Probing CSI driver for readiness
W0114 14:45:03.278921       1 metrics.go:142] metrics endpoint will not be started because `metrics-address` was not specified.
I0114 14:45:03.291063       1 controller.go:117] Register Pod informer for resizer kadalu
I0114 14:45:03.302767       1 controller.go:241] Starting external resizer kadalu
I0114 14:53:34.598992       1 event.go:282] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"pv1", UID:"10cbbc2a-8591-4db6-a6ff-11326ed4cd2a", APIVersion:"v1", ResourceVersion:"2865496", FieldPath:""}): type: 'Normal' reason: 'Resizing' External resizer is resizing volume pvc-10cbbc2a-8591-4db6-a6ff-11326ed4cd2a
I0114 14:53:37.506997       1 event.go:282] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"pv1", UID:"10cbbc2a-8591-4db6-a6ff-11326ed4cd2a", APIVersion:"v1", ResourceVersion:"2865496", FieldPath:""}): type: 'Normal' reason: 'VolumeResizeSuccessful' Resize volume succeeded
```
```
vatsa@vatsa-VirtualBox:~/kadalu$ sudo kubectl logs -nkadalu server-storage-pool-1-0-0 quotad
+ pid=0
+ cmd=/usr/bin/python3
+ script=/kadalu/server.py
+ trap 'kill ${!}; term_handler' SIGTERM
+ pid=6
+ true
+ wait 7
+ tail -f /dev/null
+ /usr/bin/python3 /kadalu/server.py
[2021-01-14 14:50:22,656] INFO [quotad - 105:handle_quota] - Quota set for size	 size=104857600
[2021-01-14 14:53:37,227] INFO [quotad - 105:handle_quota] - Quota set for size	 size=209715200
```
If the claim is lesser than the previous one or the claim is greater than the storage-pool, the claim is declined.

**Comment by vatsa287** [comment]
> @vatsa287 Please check why below steps are necessary:
> 
> > Set 'allowVolumeExpansion' to 'true' in storage class.
> > Terminate the pod using the volume.
> > Then patch the PVC.
> 
> glusterfs supports online volume expansion. I feel we can support online volume expansion, and once this PR is merged, please look at what it takes to enable `onlineVolumeExpansion` feature. Stopping the pod is not ideal for our users.

https://github.com/kubernetes/kubernetes/issues/68427. I have tested in VM so pod-restart has to done for volume-expansion to be reflected.

**Comment by vatsa287** [comment]
> @vatsa287 Please check why below steps are necessary:
> 
> > Set 'allowVolumeExpansion' to 'true' in storage class.
> > Terminate the pod using the volume.
> > Then patch the PVC.
> 
> glusterfs supports online volume expansion. I feel we can support online volume expansion, and once this PR is merged, please look at what it takes to enable `onlineVolumeExpansion` feature. Stopping the pod is not ideal for our users.

Sure! I'll take a look after exams.

File: templates/storageclass-kadalu.yaml.j2 (hunk 6 above)
Comment (vatsa287) [reply templates/storageclass-kadalu.yaml.j2:8]: I have not tested with an external-configuration yet so I don't know.

Comment (vatsa287) [reply templates/storageclass-kadalu.yaml.j2:8]: should I remove this field for external-storage class?

Comment (amarts) [reply templates/storageclass-kadalu.yaml.j2:8]: Let it be there for now!
